### PR TITLE
Send batched metrics

### PR DESCRIPTION
**Pull request by shubhitms**
- Send batched metrics
- Send batched metrics

### CODE REVIEW

File: src/memory/events.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        }
+        const events = this.batch;
+        this.batch = [];
+        try {
+            await backOff(() => this.distClient.sendFlagEvaluationMetrics(new SendFlagEvaluationMetricsRequest({
+                events,
+                sessionKey: this.sessionKey
+            })), this.backoffOptions);
+        } catch (e) {
+            console.log(`failed to send metrics batch: ${e}`);
```
Comment (shubhitms) [inline src/memory/events.ts:63]: gonna do a follow-up to do consistent logging across the board

File: src/memory/__tests__/backend.test.ts [hunk 2]
```diff
@@ -90,8 +92,56 @@ test('test backend', async () => {
...
+
+    const mockSendEvents = jest.fn();
+    Object.defineProperty(backend.distClient, "sendFlagEvaluationMetrics", { value: mockSendEvents });
+    jest.spyOn(backend.distClient, "sendFlagEvaluationMetrics").mockImplementation(async () => {
+        return new SendFlagEvaluationMetricsResponse();
+    });
 
+    await backend.initialize();
+    expect(await backend.getBoolFeature('ns-1', 'bool', new ClientContext().setBoolean('key', true))).toEqual(true);
+    expect(mockSendEvents.mock.calls.length).toEqual(1);
```
Comment (passichenko) [inline src/memory/__tests__/backend.test.ts:143]: make another evaluation and check that it sends 2 batches of metrics?

File: src/memory/__tests__/backend.test.ts [hunk 3]
```diff
@@ -90,8 +92,56 @@ test('test backend', async () => {
...
-    
+
+    const calls = mockSendEvents.mock.calls;
+    expect(calls.length).toEqual(1);
+    expect(calls[0].length).toEqual(1);
+    const sendRequest = calls[0][0] as SendFlagEvaluationMetricsRequest;
+    expect(sendRequest.events.length).toEqual(6);
+    expect(sendRequest.sessionKey).toEqual('session-key');
+    for (const event of sendRequest.events) {
+        switch (event.featureName) {
```
Comment (passichenko) [inline src/memory/__tests__/backend.test.ts:103]: nit: check other properties (maybe at least most important ones like team, ns, name, etc.)

File: src/memory/events.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            numOfAttempts: 4,
+            maxDelay: 500 // 500 ms
+        };
+    }
+
+    async init(sessionKey: string) {
+        this.sessionKey = sessionKey;
+        this.interval = setIntervalAsync(async () => {
+            await this.sendBatch();
+        }, 15 * 1000 /* 5s */);
```
Comment (passichenko) [inline src/memory/events.ts:37]: this is 15s, right?

**Review by passichenko — APPROVED** [review]

File: src/memory/__tests__/backend.test.ts (hunk 3 above)
Comment (shubhitms) [reply src/memory/__tests__/backend.test.ts:103]: good catch, i missed some checks here

File: src/memory/events.ts (hunk 4 above)
Comment (shubhitms) [reply src/memory/events.ts:37]: yup, good catch

**Pull request merged**